src/main/java/net/minecraftforge/common/data/JsonCodecProvider.java

```diff
@@ -0,0 +1,92 @@
+package net.minecraftforge.common.data;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import com.google.gson.JsonElement;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.JsonOps;
+
+import cpw.mods.modlauncher.api.LamdbaExceptionUtils;
+import net.minecraft.core.Registry;
+import net.minecraft.core.RegistryAccess;
+import net.minecraft.data.CachedOutput;
+import net.minecraft.data.DataGenerator;
+import net.minecraft.data.DataProvider;
+import net.minecraft.resources.RegistryOps;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.PackType;
+import net.minecraftforge.forge.event.lifecycle.GatherDataEvent;
+
+/**
+ * <p>Dataprovider for using a Codec to generate jsons.
+ * Path names for jsons are derived from the given registry folder and each entry's namespaced id, in the format:</p>
+ * <pre>
+ * {@code <assets/data>/entryid/registryfolder/entrypath.json }
+ * </pre>
+ * 
+ * @param <T> the type of thing being generated.
+ * @param dataGenerator DataGenerator provided by {@link GatherDataEvent}.
+ * @param dynamicOps DynamicOps to encode values to jsons with using the provided Codec, e.g. {@link JsonOps.INSTANCE}.
+ * @param packType PackType specifying whether to generate entries in assets or data.
+ * @param registryFolder String representing the registry folder, e.g. "dimension" or "cheesemod/cheese".
+ * @param codec Codec to encode values to jsons with using the provided DynamicOps.
+ * @param entries Map of named entries to serialize to jsons. Paths for values are derived from the ResourceLocation's entryid:entrypath as specified above.
+ */
+public record JsonCodecProvider<T>(DataGenerator dataGenerator, String modid, DynamicOps<JsonElement> dynamicOps, PackType packType, String registryFolder, Codec<T> codec, Map<ResourceLocation, T> entries) implements DataProvider
```

Comment:
Maybe rename registry folder to something else for more generic (non dynamic registry) use cases